cql3: expr: introduce search_and_replace()

avikivity opened the pull request:
Introduce a general-purpose search and replace function to manipulate
expressions, and use it to simplify replace_column_def() and
replace_token().

cvybhu commented on cql3/expr/expression.cc:
```diff
@@ -1265,6 +1222,67 @@ expression replace_token(const expression& expr, const column_definition* new_cd
 
+expression search_and_replace(const expression& e,
+        const noncopyable_function<std::optional<expression> (const expression& candidate)>& replace_candidate) {
+    //auto recurse = std::bind(search_and_replace, std::placeholders::_1, std::ref(replace_candidate));
```
Some commented code was left in

avikivity replied:
Doh, will fix.

cvybhu commented on cql3/expr/expression.cc:
```diff
@@ -1265,6 +1222,67 @@ expression replace_token(const expression& expr, const column_definition* new_cd
+    };
+    return replace_candidate(e).value_or(
+            std::visit(overloaded_functor{
+                [] (bool b) -> expression { return b; },
```
`bool` qualifies as a `LeafExpression`, it could be handled with other `LeafExpressions`

avikivity replied:
Yes

cvybhu commented on cql3/expr/expression.hh:
```diff
@@ -477,6 +491,10 @@ extern expression replace_column_def(const expression&, const column_definition*
 extern expression replace_token(const expression&, const column_definition*);
 
+// Reursively copies e and returns it. Calls replace_candidate() on all nodes. If it returns nullopt,
+// continue with the copying. If it returns an expression, that expression replaces the current node.
```
It would be good to mention that it replaces only LHS in `binary_operator`

cvybhu replied:
also a typo in `Reursively`

dekimir replied:
Or make it recurse on both sides?  Existing code will still work due to upstream constraints, and we could leverage the future generalized recursive visitor.

avikivity replied:
It's can't recurse the RHS because it's not an expression today. Once it is, of course.

nyh replied:
See also my comment about this above. Whatever type the rhs is (I didn't know it wasn't an expression), you can also implement recurse() for that type, even as a a no-op. Later, when rhs becomes an expressure, the right recurse() will be magically called - and you won't need to remember to fix yet another place in the code.

avikivity replied:
recurse is a named lambda. I could add a no-op recurse_term() and call it, and when term is removed, recurse_term() won't compile any more.

avikivity replied:
Unfortunately, that makes column_value and column_value_tuple non-leaf expressions too (they contain terms for subscripts). We should replace them with subscript{expression, expression} and make column_value a true leaf (and replace column_value_tuple with the existing tuple_constructor).

cvybhu commented on cql3/expr/expression.hh:
```diff
@@ -97,6 +97,20 @@ using expression = std::variant<bool, conjunction, binary_operator, column_value
+        || std::same_as<column_value_tuple, E> 
+        || std::same_as<token, E> 
+        || std::same_as<unresolved_identifier, E> 
+        || std::same_as<column_mutation_attribute, E> 
```
`column_mutation_attribute` contains a `nested_expression`

avikivity replied:
Wow, bad bug, will fix.

cvybhu reviewed (changes requested):
> It cannot be
used yet to simplify replace_token, since it should be called on the left
hand side of a binary operator, not the right hand side,

Both `replace_column_def` and `replace_token` replace on the LHS, it should be possible to implement `replace_token`.

cvybhu commented on cql3/expr/expression.cc:
```diff
@@ -1265,6 +1222,67 @@ expression replace_token(const expression& expr, const column_definition* new_cd
+    auto recurse = [&] (const expression& e) -> expression {
+        return search_and_replace(e, replace_candidate);
+    };
+    return replace_candidate(e).value_or(
```
`value_or` takes the alternative value by value. That would mean that even when `replace_candidate` returns some value, we are still going to evaluate the alternative option. This doesn't seem good for performance.

avikivity replied:
Right

dekimir commented on cql3/expr/expression.hh:
```diff
@@ -97,6 +97,20 @@ using expression = std::variant<bool, conjunction, binary_operator, column_value
 
+// An expression that doesn't contain subexpressions
+template <typename E>
+concept LeafExpression
```
Optional: if we make this a real type, there's much less risk of it getting out-of-date.  At the price of making expression less flat, of course.

avikivity replied:
This falls into the OO trap: there is one favored way to categorize expressions, and anything else has to use secondary means.

With these concepts, any subset of expression types can be made into a category. They can overlap.

dekimir replied:
OK.  We'll just have to maintain it vigilantly.

nyh replied:
If we miss one, won't the overloaded_visitor code above complain during compilation, that some of the cases aren't covered?

dekimir replied:
It will, but you could fix it by adding the branch explicitly to the visitor (instead of adding the type to LeafExpression).

dekimir commented on cql3/expr/expression.cc:
```diff
@@ -1265,6 +1222,67 @@ expression replace_token(const expression& expr, const column_definition* new_cd
+                [&] (const field_selection& fs) -> expression {
+                    return field_selection{recurse(*fs.structure), fs.field};
+                },
+                [&] (LeafExpression auto const& e) -> expression {
```
I'm surprised `auto` is tolerated here.  TIL.

avikivity replied:
It's mandatory!

avikivity replied:
Yes, except `template` cannot introduce a lambda. You could write [&] <LeafExpression L> (const L& e) -> expression, but using `LeafExpression auto` is more concise.

avikivity commented:
> > It cannot be
> > used yet to simplify replace_token, since it should be called on the left
> > hand side of a binary operator, not the right hand side,
> 
> Both `replace_column_def` and `replace_token` replace on the LHS, it should be possible to implement `replace_token`.

One day soon, the RHS of binary_expression will also be an expression. search_and_replace will then recurse into the RHS too.

What happens to the expression

     WHERE regular_column_that_is_indexed = :v1 AND token(pk) >= token(:tok)

?

Actually, I think it will be fine. token(19) will not translate into the token expression, it will translate to the token function called on bind_variable:

    binary_operator{
        token
        oper_t::GE
        function_call{
               function_name{system.token}
               bind_variable{tok}
        }
    }

I was worried it will replace both the LHS and RHS, but it won't. I'll add a patch that updates replace_token().

There's still odd things like

    WHERE regular_column_that_is_indexed = :v1 AND token(pk) >= token(pk)

but when our grammar allows it, the query planner should not allow that fragment of the WHERE clause to be used for slicing the index, only filtering.

(a first step could be to recognize `WHERE 4 < ck AND ck < 17`)

avikivity commented:
v2:
 - also rewrite replace_token()
 - update cover letter

dekimir commented:
@avikivity I count at least three comments by @cvybhu that you haven't addressed -- what am I missing?

avikivity commented:
> @avikivity I count at least three comments by @cvybhu that you haven't addressed -- what am I missing?

Maybe I missed them, I'll look again.

avikivity commented:
v3:
 - remove newly introduced commented-out code
 - search_and_replace() now lets `bool` be handled as other LeafExpression:s
 - fix typo in search_and_replace() comment.
 - mention that search_and_replace doesn't operate on binary_operator RHS
 - column_mutation_attribute is not a LeafExpression! Added code to recurse into it, and removed it
   from LeafExpression concept
 - removed unnecessary evaluation of copying action of search_and_replace when the user function
   elected to replace a subexpression

nyh reviewed (approved):
Looks good. But I was also bothered about the yet-another-place we add assuming binary operators only have an expression on the lhs. I thought it might be possible to be prepared by having a no-op recurse(term) in addition to recurse(expression) - so if rhs ever becomes a full-fledged expression you won't need to remember to fix this code. What do you think?

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -1265,6 +1265,73 @@ expression replace_token(const expression& expr, const column_definition* new_cd
+                    };
+                },
+                [&] (const binary_operator& oper) -> expression {
+                    return binary_operator(recurse(*oper.lhs), oper.op, oper.rhs);
```
We can't, because recurse() gets an expression as input.

I'm sure that @cvybhu will remember to fix this when he converts term to expressions.

nyh replied:
Ok, if you're sure this will be remembered, it's fine. I'm worried it won't be easy to remember all these places, though.
